### PR TITLE
Refactor linesearch.

### LINKED ISSUE #320: perform_linesearch_step
Lots of duplicate code in the algorithms that call linesearch. Maybe we should create a `perform_linesearch_step` function?

Ref https://github.com/JuliaOpt/Optim.jl/pull/275#issuecomment-266857553

### PR DESCRIPTION
**Pull request by pkofod**
fixes #320

- [x] include clearing
- [x] move alphaguess functionality
- [x] check dphi0 for those that need it

### CODE REVIEW

**Comment by KristofferC** [comment]
Nice code deletion

File: src/bfgs.jl [hunk 1]
```diff
@@ -79,24 +78,10 @@ function update_state!{T}(d, state::BFGSState{T}, method::BFGS)
...
-    catch ex
-        if isa(ex, LineSearches.LineSearchException)
-            lssuccess = false
-            state.f_calls, state.g_calls = state.f_calls + ex.f_update, state.g_calls + ex.g_update
-            state.alpha = ex.alpha
-            Base.warn("Linesearch failed, using alpha = $(state.alpha) and exiting optimization.")
-        else
-            rethrow(ex)
-        end
+    if method.resetalpha == true
```
Comment (pkofod) [inline src/bfgs.jl]: this should just be in the function

File: src/newton.jl [hunk 2]
```diff
@@ -69,29 +68,12 @@ function update_state!{T}(d, state::NewtonState{T}, method::Newton)
     push!(state.lsr, zero(T), state.f_x, dphi0)
 
     # Determine the distance of movement along the search line
-    try
-        if method.resetalpha == true
```
Comment (pkofod) [inline src/newton.jl]: this was deleted, but will be in `perform_linesearch` next commit

File: src/utilities/perform_linesearch.jl [hunk 3]
```diff
@@ -0,0 +1,23 @@
+perform_linesearch(state, method, d) = perform_linesearch(state, method, d, state.alpha)
+function perform_linesearch(state, method, d, alphaguess)
+    # Determine the distance of movement along the search line
+    try
+        if method.resetalpha == true
```
Comment (anriseth) [inline src/utilities/perform_linesearch.jl]: What happens if 'method' doesn't have the 'resetalpha' field?

Comment (pkofod) [inline src/utilities/perform_linesearch.jl]: Well, that is what you call a good question. It won't work. I'll fix it when at a pc.

**Comment by pkofod** [comment]
Alright, so now a linesearch step is one line in the `update_state!` calls. Right now `alphaguess` and `checked_dphi0` are in the `perform_linesearch` file, but could be in the respective method files.

**Comment by pkofod** [comment]
@anriseth when you have some time I'd be happy if you could have one last look.

**Comment by anriseth** [comment]
> @anriseth when you have some time I'd be happy if you could have one last look.

This looks good, I have no particular comments.

**Pull request merged**